### PR TITLE
feat: address validation when building txs

### PR DESCRIPTION
**Pull request by woodenfurniture**
## Description

Reimplements address validation during tx build rather than during broadcast.

- we should consider the intended lifecycle when we move tx build to the backend since the verified status is only valid during runtime (verifying it backend-side would be far more robust). We can progress this in a follow-up by building txs for trades in `src/lib/swapper/swappers/ThorchainSwapper/endpoints.ts` - UXTO is already in end-state.

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

NA

## Risk

- Risk of broken sends/trades/defi across all chains, across the whole app.
- Risk of sanctioned addresses being able to transact

## Testing

- Check all ways of transacting across the app, across all chains

### Engineering

- includes automated introspection of UXTO input addresses, please check validity and correctness of this
- Monkey-patch the sanction check and repeat testing to ensure no transactions possible

### Operations

☝️

### CODE REVIEW

**Comment by woodenfurniture** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #5454** <a href="https://app.graphite.dev/github/pr/shapeshift/web/5454" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/5454?utm_source=stack-comment).

File: src/plugins/walletConnectToDapps/utils/EIP155RequestHandlerUtil.ts [hunk 1]
```diff
@@ -141,10 +142,10 @@ export const approveEIP155Request = async ({
           to: signTransaction.to,
           value: signTransaction.value ?? convertNumberToHex(0),
           ...gasData,
-        },
+        }),
         wallet,
       })
-      return formatJsonRpcResult(id, signature)
+      return formatJsonRpcResult(id, signature.unwrap())
```
Comment (woodenfurniture) [inline src/plugins/walletConnectToDapps/utils/EIP155RequestHandlerUtil.ts:148]: note this transaction was unverfieid in the previous iteration 💀

File: src/plugins/walletConnectToDapps/utils/EIP155RequestHandlerUtil.ts [hunk 2]
```diff
@@ -125,14 +121,19 @@ export const approveEIP155Request = async ({
         ? convertNumberToHex(customTransactionData.nonce)
         : signTransaction.nonce
       if (!nonce) throw new Error('approveEIP155Request: missing nonce')
+      if (!accountNumber) throw new Error('approveEIP155Request: missing accountNumber')
```
Comment (woodenfurniture) [inline src/plugins/walletConnectToDapps/utils/EIP155RequestHandlerUtil.ts:124]:
```suggestion
      if (accountNumber === undefined) throw new Error('approveEIP155Request: missing accountNumber')
```

File: src/lib/investor/investor-foxy/api/api.ts [hunk 3]
```diff
@@ -133,46 +133,36 @@ export class FoxyApi {
...
+        ...(shouldUseEIP1559Fees ? { maxFeePerGas, maxPriorityFeePerGas } : { gasPrice }),
+      },
+      undefined,
+    )
     if (wallet.supportsOfflineSigning()) {
       const signedTx = await this.adapter.signTransaction({ txToSign, wallet })
-      if (dryRun) return signedTx
+      if (dryRun) return signedTx.unwrap()
       try {
+        // TODO: remove this as we should never be broadcasting outside the safety of chain adapters (due to address blocking)
```
Comment (woodenfurniture) [inline src/lib/investor/investor-foxy/api/api.ts:152]: do we want to action this as part of this PR?

File: src/hooks/useEvm/useEvm.ts [hunk 4]
```diff
@@ -17,11 +17,6 @@ const chainIdFromEthNetwork = (
   return supportedEvmChainIds.find(chainId => fromChainId(chainId).chainReference === ethNetwork)
 }
 
-export const getSupportedEvmChainIds = () =>
-  Array.from(getChainAdapterManager().keys()).filter(
-    chainId => fromChainId(chainId).chainNamespace === CHAIN_NAMESPACE.Evm,
-  )
-
```
Comment (woodenfurniture) [inline src/hooks/useEvm/useEvm.ts:24]: moved to utils and generalized to all namespaces for better code reuse

File: packages/chain-adapters/src/evm/EvmBaseAdapter.ts [hunk 5]
```diff
@@ -629,7 +627,7 @@ export abstract class EvmBaseAdapter<T extends EvmChainId> implements IChainAdap
         to,
         chainId: Number(fromChainId(this.chainId).chainReference),
         data,
-        nonce: numberToHex(account.chainSpecific.nonce),
+        nonce: nonce ?? numberToHex(account.chainSpecific.nonce),
```
Comment (woodenfurniture) [inline packages/chain-adapters/src/evm/EvmBaseAdapter.ts:630]: this allows us to set the nonce without circumventing the verified wrapper

File: src/plugins/walletConnectToDapps/utils/EIP155RequestHandlerUtil.ts [hunk 6]
```diff
@@ -88,31 +88,27 @@ export const approveEIP155Request = async ({
...
+      const txToSign = await chainAdapter.buildCustomTx(
+        {
+          wallet,
+          accountNumber,
+          to: sendTransaction.to,
+          data: sendTransaction.data,
+          value: sendTransaction.value ?? '0',
+          // https://docs.walletconnect.com/2.0/advanced/rpc-reference/ethereum-rpc#eth_sendtransaction
+          gasLimit: customTransactionData.gasLimit ?? sendTransaction.gasLimit ?? '90000',
+          nonce: didUserChangeNonce ? maybeAdvancedParamsNonce : undefined,
```
Comment (woodenfurniture) [inline src/plugins/walletConnectToDapps/utils/EIP155RequestHandlerUtil.ts:101]: see https://github.com/shapeshift/web/pull/5454/files#r1357568226

File: packages/chain-adapters/src/utxo/UtxoBaseAdapter.ts [hunk 7]
```diff
@@ -620,4 +608,18 @@ export abstract class UtxoBaseAdapter<T extends UtxoChainId> implements IChainAd
...
+  async verifySignTx(tx: SignTx<T>) {
+    const addresses = (
+      await Promise.all(
+        tx.inputs.map(async input => {
+          const tx = await this.providers.http.getTransaction(input)
+          // assumes input.vout is an index to tx.vout[]
+          return tx.vout[input.vout].addresses ?? []
+        }),
+      )
+    ).flat()
```
Comment (woodenfurniture) [inline packages/chain-adapters/src/utxo/UtxoBaseAdapter.ts:621]: would appreciate input on this - is it correct?

File: packages/chain-adapters/src/evm/EvmBaseAdapter.ts [hunk 8]
```diff
@@ -698,4 +703,22 @@ export abstract class EvmBaseAdapter<T extends EvmChainId> implements IChainAdap
...
+  ) {
+    // TODO: simulate the tx to introspect contract interactions for detailed address list
+    const addresses = [
+      await this.getAddress({ wallet, accountNumber }),
+      tx.to,
+      receiveAddress,
+    ].filter((address): address is string => address !== undefined)
+
+    return verify(addresses, tx)
+  }
```
Comment (woodenfurniture) [inline packages/chain-adapters/src/evm/EvmBaseAdapter.ts:723]: this is probably the most fragile of the whole PR, but no worse than whats in develop currently - it requires us to manually pass in the correct addresses. Later we'll want to simulate the tx and verify those to automate it

File: packages/chain-adapters/src/utils/verify.ts [hunk 9]
```diff
@@ -0,0 +1,72 @@
+import assert from 'assert'
+import axios from 'axios'
+
+// do not export this to prevent misuse
+const VERIFIED_HANDLE: unique symbol = Symbol('verifiedHandle')
+
+export interface Verified<T> {
+  [VERIFIED_HANDLE]: true
+  unwrap: () => T
+}
```
Comment (gomesalexandre) [inline packages/chain-adapters/src/utils/verify.ts:10]: 🧠

File: src/lib/utils/index.ts [hunk 10]
```diff
@@ -220,3 +221,15 @@ export const timeout = <Left, Right>(
     ),
   ])
 }
+
+export const getSupportedChainIds = () => {
```
Comment (gomesalexandre) [inline src/lib/utils/index.ts:225]:
```suggestion
export const getSupportedChainIdsByChainNamespace = () => {
```

File: packages/chain-adapters/src/utxo/UtxoBaseAdapter.ts [hunk 11]
```diff
@@ -620,4 +608,18 @@ export abstract class UtxoBaseAdapter<T extends UtxoChainId> implements IChainAd
 
     return publicKeys[0]
   }
+
+  async verifySignTx(tx: SignTx<T>) {
+    const addresses = (
+      await Promise.all(
+        tx.inputs.map(async input => {
+          const tx = await this.providers.http.getTransaction(input)
```
Comment (gomesalexandre) [inline packages/chain-adapters/src/utxo/UtxoBaseAdapter.ts:616]: We should be able to introspect inputs and outputs from the build UTXOs themselves, no need to fetch anything here

File: packages/chain-adapters/src/utxo/UtxoBaseAdapter.ts (hunk 7 above)
Comment (gomesalexandre) [reply packages/chain-adapters/src/utxo/UtxoBaseAdapter.ts:621]: We will want to verify both inputs and outputs UTXOs here @woodenfurniture meaning no UTXO is consumed with a sanctioned address, as well as none of the outputs being to a sanctioned address.

tl;dr of how things work is we're augmenting `coinselect` in our `utxoSelect` implementation, and then again this augmented version is augmented in `signTxInputs` / `signTxOutputs`, bringing in additonal metadata.

Ideally, `utxoSelect` itself would be the one doing those checks so we don't have to deep-introspect the `SignTx` type here, though that would get bloaty.
Preferrably, if we don't want to bloat `utxoSelect` with this, we can do the verification logic on the destructured `inputs` and `outputs` of `coinSelectResult`. This will give us direct access to `input.address` for each input.

File: packages/chain-adapters/src/utxo/UtxoBaseAdapter.ts (hunk 11 above)
Comment (woodenfurniture) [reply packages/chain-adapters/src/utxo/UtxoBaseAdapter.ts:616]: perhaps we can pile on this in your am to discuss?

**Comment by woodenfurniture** [comment]
noting this PR is awaiting more detail on the architecture of the SDK before merge and/or rework

**Comment by 0xean** [comment]
@woodenfurniture to bring over useful bits in separate PR.
